mkosi/kmod.py

```diff
@@ -172,30 +172,22 @@ def gen_required_kernel_modules(
         mods = set((modulesd / "kernel").rglob("*.ko*"))
         firmware = set()
 
-    yield modulesd.parent
-    yield modulesd
-    yield modulesd / "kernel"
-
-    if (root / "usr/lib/firmware").exists():
-        yield root / "usr/lib/firmware"
-
-    for d in (modulesd, root / "usr/lib/firmware"):
-        for p in (root / d).rglob("*"):
-            if p.is_dir():
-                yield p
-
-    for p in sorted(mods) + sorted(firmware):
-        yield p
-
-    for p in (root / modulesd).iterdir():
-        if p.name.startswith("modules"):
-            yield p
+    # We do not want to yield directories lower than /usr/lib, especially none outside of our root namespace.
+    # Therefore we disregard the lowest few parent directores.
+    # This also improves performance as fewer set memberships have to be checked.
+    prefix_length = len((root / "usr/lib").parts)
+    required_paths = mods | firmware | set(
```

Comment:
Please just use the set comprehension: `{ parent for … }`

Reply:
The whole thing could also be written as
```python
firmwared = root / "usr/lib/firmware"
required_paths = mods | firmware | { 
    parent
    for file in mods | firmware
    for parent in file.parents
    if parent.is_relative_to(modulesd) or parent.is_relative_to(firmwared)
}
```
It's a bit clearer (no list index juggling) and performs the same. It's also a bit tighter in what it accepts.